l2geth/core/types/receipt.go

```diff
@@ -352,7 +354,20 @@ func (r Receipts) DeriveFields(config *params.ChainConfig, hash common.Hash, num
 		if txs[i].To() == nil {
 			// Deriving the signer is expensive, only do if it's actually needed
 			from, _ := Sender(signer, txs[i])
-			r[i].ContractAddress = crypto.CreateAddress(from, txs[i].Nonce())
+			nonce := txs[i].Nonce()
+
+			if rcfg.UsingOVM {
+				sysAddress := rcfg.SystemAddressFor(config.ChainID, from)
+				// If nonce is zero, and the deployer is a system address deployer,
+				// set the provided system contract address.
+				if sysAddress != rcfg.ZeroSystemAddress && nonce == 0 {
```

Comment:
This will be inconsistent with the address that the contract is actually deployed at in the case `len(codesize) > 0`

Reply:
Updated to detect a contract creation transaction at this nonce by adding a check for `tx[i].To() == nil`. The only possible address for a contract created at nonce 0 from a system address deployer is the system address.

Reply:
I took at look at the rest of `evm.go` - there's a check that will revert if the codesize is > 0:

```go
	// Ensure there's no existing contract already at the designated address
	contractHash := evm.StateDB.GetCodeHash(address)
	if evm.StateDB.GetNonce(address) != 0 || (contractHash != (common.Hash{}) && contractHash != emptyCodeHash) {
		return
```

So, I removed the additional check in the line above.